### PR TITLE
Limit IndexRequest toString() length

### PR DESCRIPTION
**Pull request by sabi0**
This patch significantly reduces memory usage when indexing large documents.

Apparently the whole input document (`source`) is rendered into `IndexRequest` `toString()` result.
And two distinct copies of such `String`s are then used as `ReplicationTask` descriptions.
(in `TransportReplicationAction.PrimaryPhase` and `TransportReplicationAction.ReroutePhase`)

Thus, for 10 MB input in UTF-8 there would be about 40 MB of 'excessively allocated' memory.

The patch changes `IndexRequest.toString()` implementation so that only piece of large `source` is used.

### CODE REVIEW

**Comment by clintongormley** [comment]
(CLA was signed but the CLA checker is down atm)

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 1]
```diff
@@ -549,8 +558,17 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         String sSource = "_na_";
         try {
-            sSource = XContentHelper.convertToJson(source, false);
-        } catch (Exception e) {
+            if (source.length() > MAX_SOURCE_LENGTH_IN_TOSTRING) {
+                // We might slice in the middle of a UTF-8 character here, but BytesRef.utf8ToString() will read more bytes if needed.
+                // There is no risk of ArrayIndexOutOfBoundsException either as source is assumed to be properly UTF-8 encoded.
+                sSource = source.slice(0, MAX_SOURCE_LENGTH_IN_TOSTRING).utf8ToString() + "..." +
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I think `convertToJson` was important here. If the document was submitted in cbor or smile this will blow up.

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 2]
```diff
@@ -549,8 +558,17 @@ public void writeTo(StreamOutput out) throws IOException {
...
+                    new ByteSizeValue(source.length()).toString();
+            }
+            else {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: nit: move this to the line above.

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 3]
```diff
@@ -549,8 +558,17 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         String sSource = "_na_";
         try {
-            sSource = XContentHelper.convertToJson(source, false);
-        } catch (Exception e) {
+            if (source.length() > MAX_SOURCE_LENGTH_IN_TOSTRING) {
+                // We might slice in the middle of a UTF-8 character here, but BytesRef.utf8ToString() will read more bytes if needed.
+                // There is no risk of ArrayIndexOutOfBoundsException either as source is assumed to be properly UTF-8 encoded.
+                sSource = source.slice(0, MAX_SOURCE_LENGTH_IN_TOSTRING).utf8ToString() + "..." +
+                    new ByteSizeValue(source.length()).toString();
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I like including the original size here. Maybe instead if the source is chopped it should read like `first 2048 characters out of 10122123: _slice_of_the_source_`.

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by nik9000** [comment]
I've requested a few changes but I think the idea is sound.

**Comment by nik9000** [comment]
After reading this, I also filed #22866.

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 1 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I feel like we might get a working solution by adding something like `XContentHelper.convertToJsonFragment(source, maxFragmentSize);` that would construct a XContentBuilder by passing it a modified `BytesStreamOutput` that would through an exception when it reaches a certain size, then we can intercept this exception add "..." at the end and return it as a string.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: we should be careful here and check for sources that are binary (SMILE etc..)

**Comment by bleskes** [comment]
I'm fine with making the toString() smarter and not create 100MB strings, but I'm not sure this is the right fix for the task description issue. It's a big shame that we have to parse json just to produce a task description. IMO we should use the index name, type and doc ID and be happy. Also note that all of this is going away (I believe) as we move to bulks with a single item.

**Comment by bleskes** [comment]
I would make org.elasticsearch.action.support.replication.ReplicationRequest#getDescription abstract and force people to come up with a sane description on a per request basis. IMO `toString` is associated too much with logging and debugging.

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 1 above)
Comment (sabi0) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: You are right. It was shortsighted of me to assume `IndexRequest` would always have source in JSON. I'll refactor the code to take care of other formats too. This might take a few days though as I don't have much spare time at the moment.

P.S. I have problem setting up working environment with IntelliJ IDEA. Suggested `gradle idea` + 'Create Project from Sources' approach does not quite work (e.g. the module lacks any dependencies). Is there more up to date guide? Or would anyone be willing to share the ready-made IDEA project?

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 3 above)
Comment (sabi0) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I can change that, no problem.
The only concern is that we don't know the total number of 'characters' as the `source` length is in bytes. That makes the message not very accurate.

How about: `Source is too big (10122123 bytes), starts with: _slice_goes_here_...`?
Also do you think it should print the exact number of bytes or the shorter '9.7mb' will be ok?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I don't think we need the exact number of bytes and if bytes is what we have we should use it. No reason to work hard to get characters.

**Comment by nik9000** [comment]
> I would make org.elasticsearch.action.support.replication.ReplicationRequest#getDescription abstract and force people to come up with a sane description on a per request basis.

I'm beginning to agree with you. I think it'd solve #22866 as well by forcing us to think every time. @imotov, what do you think?

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: @bleskes was talking below about using only the index, source, and id for the task description. The toString change would still be nice, but less important then. That seems like a good choice to me.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I don't use idea but I think lots of others do. They should be able to comment on this. I think those steps are supposed to work.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: >  talking below about using only the index, source, and id

probably a typo, but just in case - index, type and id (not source)

**Comment by nik9000** [comment]
Right! Sorry. My mistake.

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 1 above)
Comment (sabi0) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: FYI: I figured out the IDEA issue. It should actually be `gradle idea` + 'File > Open...' rather than 'File > New > Project from Existing Sources...'
Gradle command creates all the project files. And you don't want to re-create them with IDEA.

It might be worth updating the corresponding section in the /CONTRIBUTING.md

Comment (sabi0) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: @imotov
The problem with throwing an exception from modified `OutputStream` is that we might stop in the middle of UTF-8 sequence. And then we'd get either `ArrayIndexOutOfBounds` in `UnicodeUtil.UTF8toUTF16(...)` or some unspecified behavior in `new String(data, "UTF-8")`.

There is a workaround: collect `fragmentLength + 3` bytes, but only use `fragmentLength` bytes for decoding with `UnicodeUtil` (which would then read more bytes if needed). But this would be quite an ugly hack.

Another option is to print CBOR, SMILE, etc. bytes in hex, a-la "03 1A B7 ..." instead of converting them to JSON. Or just omit any non-JSON source altogether. This would require guessing the content type though.

Maybe we should then just skip any large source? And only print the message like "Source is too big - 5kb". Having the excerpt however might be quite useful for debugging.

Comment (imotov) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: @sabi0 I was thinking that if you override writeBytes instead of writeByte and ignore entire content of the method it should work, but it's a moot point now because I think @bleskes's idea of not including source at all is much better.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: > And only print the message like "Source is too big - 5kb"

+1 to that. Keep it simple

**Comment by sabi0** [comment]
Should I create a new PR for making `getDescription()` abstract or do that within this one?

**Comment by nik9000** [comment]
> Should I create a new PR for making getDescription() abstract or do that within this one?

I'd keep this about the toString change. Making `getDescription` abstract is likely to end up much larger than this change so lets get the fix in here.

**Comment by sabi0** [comment]
I've updated the code according to the suggestions above.

**Comment by sabi0** [comment]
.. and now there is a conflict.
Am I expected to do anything about this? (e.g. rebase the branch on the latest master)

**Comment by jasontedor** [comment]
Merge master into your branch and resolve the conflict during the merge.

**Comment by sabi0** [comment]
Hmm.. I didn't expect all those commits to appear here. Did I do something wrong?
(sorry, not so much GitHub experience here)

**Comment by s1monw** [comment]
@sabi0 you have to rebase your branch against master I think you branched off another branch on your end

**Comment by sabi0** [comment]
Not really. Here is what I did:
`git checkout master`
`git fetch upstream master`
`git merge upstream/master`
`git checkout index-request-tostring`
`git rebase master`
There were merge conflicts here. I fixed them and then ran `git add ...` followed by `git rebase --continue`.
Then I tried to push the branch to origin but got a message that changes were rejected because the HEAD is behind the remote. So I did `git pull origin index-request-tostring` to integrate the remote changes. (why would there be changes in the branch only I update?)
There were some conflicts again, so I applied the `git add ...` + `git commit ...` combo.
Then I successfully pushed the branch with `git push origin index-request-tostring` and ended up with all those extra commits appearing here.

What should I do now? Create a new clean PR?

**Comment by s1monw** [comment]
when you push to your remote, since you rebased on master locally you have to force push `-f` to your remote `git push -f origin index-request-tostring` then it will not complain and you should be good

**Comment by sabi0** [comment]
Right. Thanks.
But for this PR it's too late I suppose? (since I've already committed the unnecessary "merge conflicts")
Is there some 'git-fu' to salvage this branch? Or should I rather abandon it and create a new clean PR?

**Comment by s1monw** [comment]
> But for this PR it's too late I suppose? (since I've already committed the unnecessary "merge conflicts")

this PR should pick it up if you do it... just give it a go...

**Comment by sabi0** [comment]
But I've already pushed all the changes. (after pulling, resolving conflicts and committing)
And that's what caused all these commits appearing above.
When I try to force push now git just tells that "Everything up-to-date".

**Comment by s1monw** [comment]
```
git checkout master
git fetch upstream master
git merge upstream/master
git checkout index-request-tostring
git rebase master

git push -f origin git checkout index-request-tostring
```

is this what you did? I think you did an extra `git pull origin index-request-tostring` which is not necessary or rather messes everything up.

**Comment by s1monw** [comment]
here is what I'd do in your case:

```
git checkout master
git fetch upstream master
git reset --hard upstream/master
git checkout index-request-tostring
git rebase master

git push -f origin git checkout index-request-tostring
```

**Comment by sabi0** [comment]
Indeed. I described the full commands sequence a few comments above.
And there was `git pull origin index-request-tostring` there.

I'll now try your `reset --hard` solution, thanks!

**Comment by sabi0** [comment]
And it's fixed!
Thank you a lot!

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 4]
```diff
@@ -606,7 +615,11 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         String sSource = "_na_";
         try {
-            sSource = XContentHelper.convertToJson(source, false);
+            if (source.length() > MAX_SOURCE_LENGTH_IN_TOSTRING) {
+                sSource = "too big: " + ByteSizeValue.formatBytesSizeValue(source.length());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: what about `"Source is too big, max length [2048b] but was [XXX]"`

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java [hunk 5]
```diff
@@ -107,23 +107,26 @@ public double getPbFrac() {
 
     @Override
     public String toString() {
-        long bytes = getBytes();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: can you revert these changes and rather create a new BytesSizeValue from the source length?

**Review by s1monw — CHANGES_REQUESTED** [review]
I left some comments

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java (hunk 5 above)
Comment (sabi0) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Ok.
I though avoiding extra allocations just to call `toString()` would be better.
Static `parseBytesSizeValue` invites for the counterpart `formatBytesSizeValue`.

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 4 above)
Comment (sabi0) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I'd rather not repeat the "source" in the text. How about this format:
`... source[too big: 12345b, max printed length: 2048b]`
?

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java (hunk 5 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: just keep the change simple that extra allocation

File: core/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 4 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: then do something like this `source[n/a max source size: ...`

Comment (sabi0) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: nik9000 commented above in favor of showing the actual source length ("I like including the original size here.")

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: yeah I used `...` so explicitly `source[n/a actual source size: [12345b], max source length: 2048b]`

Comment (sabi0) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: For perspective, the complete `toString()` output will look like this:
`index {[index][type][null], source[n/a, actual length: [12.7kb], max length: 2kb]}`

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: ++

**Comment by sabi0** [comment]
Changed as requested

**Comment by s1monw** [comment]
@elasticmachine ok to test

**Review by s1monw — APPROVED** [review]

**Comment by s1monw** [comment]
@nik9000 is this ok for you to pull in?

**Comment by nik9000** [comment]
> @nik9000 is this ok for you to pull in?

I'll have another look and pull it in, sure!

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by nik9000** [comment]
> I'll have another look and pull it in, sure!

Sorry for the delay! I've been super distracted today. I'm testing locally now just to be double sure.

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
OK. I spent all afternoon trying to get the build green before trying this and failing. I'm done for the night. I'll look again in the morning.

**Comment by nik9000** [comment]
Thanks for working with us for so long @sabi0! I've merged to squashed and merged to master and now I'll backport to the 5.x branch. This should be released with Elasticsearch 5.4.

master: 09c7c5c82fdd5ce7c133de0379b3b44cf36061f9
5.x:
